lib/client.js

```diff
@@ -1199,6 +1203,334 @@ const PreClient = crypto => {
+     */
+    static async internalReadNote(authenticator, encryptionKeyPair, noteId, apiUrl) {
+      let request = await authenticator.tsv1Fetch(
+        apiUrl + `/v2/storage/notes?note_id=${noteId}`,
+        {
+          method: 'GET'
+        }
+      )
+      let storedNote = await checkStatus(request)
+      const noteJson = await storedNote.json()
+      return Client.decryptNoteJson(
+        noteJson,
+        authenticator.publicSigningKey,
+        encryptionKeyPair.privateKey,
+        encryptionKeyPair.publicKey
+      )
+    }
+
+    /*
+     * InternalReadNoteByName is an internal method used by readNoteByName
+     */
+    static async internalReadNoteByName(
+      authenticator,
+      encryptionKeyPair,
+      noteName,
+      apiUrl
+    ) {
+      let request = await authenticator.tsv1Fetch(
+        apiUrl + `/v2/storage/notes?id_string=${noteName}`,
+        {
+          method: 'GET'
+        }
+      )
+      let storedNote = await checkStatus(request)
+      const noteJson = await storedNote.json()
+      return Client.decryptNoteJson(
+        noteJson,
+        authenticator.publicSigningKey,
+        encryptionKeyPair.privateKey,
+        encryptionKeyPair.publicKey
+      )
+    }
+
+    /*
+     * InternalWriteNote is an internal method used by writeNote
+     */
+    static async internalWriteNote(
+      authenticator,
+      type,
+      data,
+      plainMeta,
+      fileMeta,
+      recipientSigningKey,
+      encryptionKeyPair,
+      options,
+      apiUrl
+    ) {
+      const accessKey = await crypto.randomKey()
+      const encryptedAccessKey = await crypto.encryptAk(
+        encryptionKeyPair.privateKey,
+        accessKey,
+        encryptionKeyPair.publicKey
+      )
+      var coreNote = new CoreNote(
+        type,
+        data,
+        plainMeta,
+        fileMeta,
+        crypto.mode,
+        recipientSigningKey,
+        authenticator.publicSigningKey,
+        encryptionKeyPair.publicKey,
+        encryptedAccessKey
+      )
+      var noteOptions = NoteOptions.decode(options)
+      let signature = await crypto.signDocument(coreNote, authenticator.privateSigningKey)
+      var note = new Note(coreNote, signature, noteOptions)
+      let encryptedNote = await crypto.encryptNote(note, accessKey)
+      let response = await authenticator.tsv1Fetch(
+        apiUrl + '/v2/storage/notes',
+        {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json'
+          },
+          body: encryptedNote.toJson()
+        },
+        noteOptions.clientId
+      )
+      let storedNoteResp = await checkStatus(response)
+      let noteJson = await storedNoteResp.json()
+      const decodedServerNote = Note.decode(noteJson)
+      return decodedServerNote
+    }
+
+    /*
+     * InternalDeleteNote is an internal method used by deleteNote
+     */
+    static async internalDeleteNote(authenticator, noteId, apiUrl) {
```

Comment:
can all of these notes related methods be broken out into a separate file? If so, should they ( a question you with fresh experience working in this code base can answer better than me) ?

Reply:
I don't know if JS allows you to extend a class to another file (for example swift allows 'extensions', to extend a class from different file). I would like them to live in a reachable place/close to the using methods since they're doing the actionable calls, but I do think it makes the object messy/ugly so I'm open to any improvements.

Reply:
not easily -- you can since at the end of the day JS is prototype based and classes are just syntactic sugar over that, but it would not look very nice or consistent.